pcap/pcap.go

```diff
@@ -0,0 +1,234 @@
+// format.  See http://wiki.wireshark.org/Development/LibpcapFileFormat
+// for information on the file format.
+//
+// We currenty read v2.4 file format with nanosecond and microsecdond
+// timestamp resolution in little-endian and big-endian encoding.
+//
+// If the PCAP data is gzip compressed it is transparently uncompressed
+// by wrapping the given io.Reader with a gzip.Reader.
+type Reader struct {
+	*peeker.Reader
+	byteOrder      binary.ByteOrder
+	nanoSecsFactor uint32
+	versionMajor   uint16
+	versionMinor   uint16
+	// timezone
+	// sigfigs
+	snaplen  uint32
+	linkType layers.LinkType
+	Offset   uint64
+	header   []byte
+}
+
+//XXX from ngwrite
+const magicMicroseconds = 0xA1B2C3D4
+const versionMajor = 2
+const versionMinor = 4
+
+const magicNanoseconds = 0xA1B23C4D
+const magicMicrosecondsBigendian = 0xD4C3B2A1
+const magicNanosecondsBigendian = 0x4D3CB2A1
+
+const magicGzip1 = 0x1f
+const magicGzip2 = 0x8b
+
+// NewReader returns a new reader object, for reading packet data from
+// the given reader. The reader must be open and header data is
+// read from it at this point.
+// If the file format is not supported an error is returned
+//
+//  // Create new reader:
+//  f, _ := os.Open("/tmp/file.pcap")
+//  defer f.Close()
+//  r, err := NewReader(f)
+//  data, ci, err := r.ReadPacketData()
+func NewReader(r io.Reader) (*Reader, error) {
+	reader := &Reader{
+		Reader: peeker.NewReader(r, 32*1024, 1024*1024),
+	}
+	if err := reader.readHeader(); err != nil {
+		return nil, err
+	}
+	return reader, nil
+}
+
+const (
+	fileHeaderLen   = 24
+	packetHeaderLen = 16
+)
+
+func (r *Reader) readHeader() error {
+	hdr, err := r.Reader.Read(fileHeaderLen)
+	if err != nil {
+		return err
+	}
+	r.header = make([]byte, fileHeaderLen)
+	copy(r.header, hdr)
+	r.Offset += fileHeaderLen
+	if magic := binary.LittleEndian.Uint32(hdr[0:4]); magic == magicNanoseconds {
+		r.byteOrder = binary.LittleEndian
+		r.nanoSecsFactor = 1
+	} else if magic == magicNanosecondsBigendian {
+		r.byteOrder = binary.BigEndian
+		r.nanoSecsFactor = 1
+	} else if magic == magicMicroseconds {
+		r.byteOrder = binary.LittleEndian
+		r.nanoSecsFactor = 1000
+	} else if magic == magicMicrosecondsBigendian {
+		r.byteOrder = binary.BigEndian
+		r.nanoSecsFactor = 1000
+	} else {
+		return fmt.Errorf("Unknown magic %x", magic)
+	}
+	if r.versionMajor = r.byteOrder.Uint16(hdr[4:6]); r.versionMajor != versionMajor {
+		return fmt.Errorf("Unknown major version %d", r.versionMajor)
+	}
+	if r.versionMinor = r.byteOrder.Uint16(hdr[6:8]); r.versionMinor != versionMinor {
+		return fmt.Errorf("Unknown minor version %d", r.versionMinor)
+	}
+	// ignore timezone 8:12 and sigfigs 12:16
+	r.snaplen = r.byteOrder.Uint32(hdr[16:20])
+	r.linkType = layers.LinkType(r.byteOrder.Uint32(hdr[20:24]))
+	return nil
+}
+
+func (r *Reader) ReadBlock(span nano.Span) ([]byte, error) {
+	header := r.header
+	if header != nil {
+		r.header = nil
+		return header, nil
+	}
```

Comment:
Seems odd that `ReadBlock` that takes a `nano.Span` as an argument would return a non-packet block. Why not have this not return the header?

Reply:
In a subsequent PR.  When I had the ng-reader I will probably introduced an interface and clean this up.